### ISSUE TITLE
Enable dry runs with new 3.0 API

### ISSUE THREAD

**Issue by vyasr** [glotzerlab/hoomd-blue#506]
## Description

Currently many users test complex HOOMD scripts by changing all the `hoomd.run` commands to just run for one step to verify that there are no errors in configuring later parts of the script. Failing to do these kinds of tests can lead to running unnecessarily long simulations before discovering a failure point later in the script. It would be nice to have a way to check that a script won't fail in this manner by executing a dry run that doesn't actually run the simulation (credit for the suggestion goes to @mphoward).

## Proposed solution

One way we could implement this would be setting some `dry-run` flag (not sure exactly how given that we're moving away from command line options, so that's an open question) that would execute the script as normal except that calls to `hoomd.run` would check for this flag and automatically execute a dry run by preparing everything up to the point of actually running the timesteps in C++. Even without initializing the C++ side it should be possible to verify some part of the correctness by taking advantage of the lazy evaluation model in 3.0 to check the parts that don't require the C++ side to be initialized.

## Developer

Implementing this feature in the most natural way requires most of the 3.0 API to be ready, so we can revisit a possible developer once that project is closer to completion.

**Comment by joaander** [comment]
What is the primary goal of this? That the user script can run to completion with no syntax errors, generate all output files, etc...? Or just that all parameters necessary for `run()` to start are set?

**Comment by mphoward** [comment]
I had in mind mostly the former as a convenience feature (checking syntax, outputs, etc.) to avoid that sad moment, e.g., when the gsd restart file fails to write after a run command because of some error. It seemed like it might be compatible with the lazy initialization Vyas described. Right now, I achieve the same by replacing all run(steps) commands by run(1).

**Comment by joaander** [comment]
In that case, does a dry run need to ensure that all scheduled updaters/analyzers/etc... execute at least once? When they do execute, should they fail to produce output files when the dry run option is active?

**Comment by mphoward** [comment]
Hmm, good question. I think probably no files should be generated, as my usual workflow is: replace all runs by run(1), run the script, look at HOOMD’s output to make sure it exited cleanly, then delete any files that were created (e.g., to avoid appending).

I’m not sure whether the objects actually need to run. It might be enough to proceed through C++ initialization, but no actual execution. I use a dry run mainly as a syntax check when invoking additional commands after a run, but I defer correctness of the object execution until the actual run, where these are usually easier to spot.

**Comment by joaander** [comment]
Excellent. Thanks for the suggestion. I think this feature will be useful and relatively easy to implement as a property of the simulation object.
```
s = Simulation()
s.dry_run = True
s.ops.add(my_operations)
s.run(1e9)  # only runs 1 step
```

At the `schedule()` time any analyzers that produce output can inspect the `dry_run` property and pass it down to C++ so they can skip file writing accordingly.

The simplest implementation would only allow `dry_run` to be modified before scheduling. We could relax that, but it would require that we keep track of all the operations that need to know about the `dry_run` setting and pass changes to them. I'm not sure we even want to allow changing `dry_run` in the middle of a simulation.

**Comment by mphoward** [comment]
I think `dry_run` should be read-only after scheduling, as it doesn't make any sense to me to switch from a dry run to a real run part way through a script. Either it could be a modifiable property prior to scheduling, or it could be an argument to the `Simulation` constructor (default `False`). Having it in one place to switch between dry and real run is perfect though!

**Comment by joaander** [comment]
Coming back to this issue since it was mentioned. HOOMD v3 currently validates all input parameters when set. Most of that validation occurs in Python, but some of it occurs in C++ on a call to schedule. Do we need a dry_run type flag aside from schedule? After scheduling, all operations passed to the simulation are initialized in C++ and ready for `run` to start. All the information necessary for run metadata, citation information ,etc... is available at this time.

@mphoward You mentioned the possibility of a GSD restart failing to write after a run. In the new API, this looks like:
```
sim = setup_simulation()
sim.run(N)
hoomd.dump.GSD.write('filename.gsd')
```
If there were a `dry_run` property set on this simulation what should this script do? Would `GSD.write` write any output? 

There are a lot of other corner cases to consider as well. How many timesteps would `run` perform? What would be the value of `sim.timestep` after the run? Would any of the operations added to the simulation be triggered? Should accessing particle state data and snapshots be possible while in a dry run? Would other immediate APIs like `update.BoxResize.update` change the system state?

**Comment by vyasr** [comment]
Before discussing feasibility, which obviously could be a significant barrier to certain choices we might want to make, I think it would be helpful to come to a consensus on how a dry run _should_ behave in an ideal world. We can prune back from there if there are certain expectations that would be very technically difficult to satisfy. From that perspective, IMO the ideal case would be for a dry run to behave exactly like a real run except that no integration would occur and no side effects would be visible. By side effects, I mean any changes that persist beyond the current Python process. In particular, this means:

1. No files would be created by HOOMD commands, but users would be responsible for handling this appropriately if they run anything else that might affect output (e.g. a `signac` job document modification, or with respect to #748, `duecredit` citation generation).
1. All operations should be triggered to make sure that they work.
1. All C++ objects should be created and their data should be accessible, and modifications (for instance of sim.state) should be reflected in the underlying objects. While enabling this might slow things down a little bit, I think the main thing we want to avoid is a situation where we wait 24 hours for a simulation to run before finding that the last dump fails (@mphoward's example), not waiting 20 seconds for a naive snapshot initialization protocol to complete (especially since that is on the user to optimize). The overhead for creating C++ objects shouldn't be too heavy IMO.
1. All variants should immediately track to their final state.

Open questions (to me):
1. The `timestep` question is a good one. My first instinct is to say that we should spoof the timestep and advance it to the end of the run, but have all operations triggered immediately and not actually integrate forward at all (so the real timestep would still be 0).
1. It may make sense to always choose a CPU device for a dry run for simplicity, especially if one use case would be to generate citations. On the other hand, one usage of a dry run would be to make sure that all the features being used in the script are supported on the proposed device: for instance, HPMC external fields don't run on the GPU. One way around this might be to catch certain failures like that and throw warnings instead of raising exceptions.

Anyone on reading this thread should feel free to edit this comment and add to these lists directly if you feel I'm missing something.

**Comment by mphoward** [comment]
My main thought behind a "dry run" option was to give a crude form of static syntax checking, and not that the outputs would be meaningful. This is important for multistage workflows where the simulation will be modified in between `run` commands, which is essentially a long barrier in the script.

Here's a pseudoexample (typical to my usage) where this feature would be really helpful:
```py
# equilibration to remove overlaps
sim = setup_simulation()
sim.run(1e6)

# make a change to particles and do a second stage of equilibration
modify_particles(sim)
sim.run(1e6)

# production, when we start saving output
attach_production_outputs(sim)
sim.run(1e7)

# just remembered I should save a restart file, let me add that to my script quickly then submit the job
hoomd.dump.GSD.write('filename.gsd', dymanic=['property'])
# oh no, I had to wait 3 days to realize I misspelled dynamic!
```
To check this code, I would usually grep for all the lines with `run(N)`, comment them out and add `run(1)`, then run the script to see that it gets to the end without error. Otherwise, if I typoed in `modify_particles`, `attach_production_outputs`, or the restart dump, I might not figure this out until several hours later. After I do this, I then go back to remove `run(1)` and uncomment `run(N)`. Then, I remove any output files and submit the script for real. I've definitely messed up one or both directions of this commenting process before though, and this wastes time.

I think the ideal behavior to emulate this procedure would be to let the user set `dry_run` equal to the number of steps they want to use in place of the run commands in the script. If possible with the current design, this could even be zero steps so that the run does not execute at all. When `sim.run(N)` is called, `N` should be ignored and replaced by `dry_run` steps. The user can choose `dry_run` to be whatever makes sense for their testing purposes (maybe that don't need it to step at all, maybe they need it to hit the first period of some execution, and what is "long" or "short" depends on the simulation). For more complex workflows, say where a different number of steps needs to be performed to test each call to `run`, `dry_run` could be specified as a list of steps to use in each.

By default `dry_run` would be unset or set to `False`, `None`, `-1`... something that indicates that the real run commands should be used. To use (or not) this checking feature, the user then just needs to add / remove one single line in the script and not change anything else.

For me, I would want `dry_run` to mock all the runs with a significantly smaller number of steps than would be needed to properly sample the simulation, and I want to be able to enable this with an easy one-line change. I think all of the outputs, etc. could still trigger as if it were in the real simulation. The user could choose how many steps to use in the mock so that it tests what they need.

**Comment by mphoward** [comment]
@vyasr and I commented at the same time, apparently, so consider this my addendum to his list. :-)

**Comment by vyasr** [comment]
:) nice timing. I like @mphoward's suggestion to literally treat dry runs as running with fewer steps instead of my suggestion to treat it as `run(0)`. It sounds like we're mostly in agreement on the other expected behaviors, and I think running a nonzero number of steps might save us from handling a bunch of edge cases that could arise from doing a `run(0)`. I'm not sure how to handle the file output case. I think the most natural option is what we've generally been saying, which is that no files should be dumped etc, but I definitely see where Mike's coming from regarding just dumping output like the script normally would, and I'm sure that as soon as we decided otherwise somebody will say "but I want to check my first GSD frame and make sure the shapes are right" or something. So maybe we go with the simplest implementation first (just do a shorter run but otherwise treat everything as normal) and make sure that satisfies everyone's use cases.

If we wanted to, we could add additional configuration options to the simulation to handle such things if we wanted, for instance by introducing a `sim.dry_run_options` namespace (basically a dict) to handle things like multiple `dry_run` number of steps like in Mike's last comment or whether or not files should be dumped. Then we could choose some defaults that we like but permit the user to configure as needed. I'm not sure if that's overkill though; I iteratively modify my scripts enough that I can imagine changing followed by dry running multiple times, but I don't know if most people do that enough to support configurability beyond a simple boolean flag.

**Comment by joaander** [comment]
I see this as a useful option for certain types of scripts but I don't see how we can possibly make it work for all users scripts. HOOMD is an API, not a black box tool following a fixed recipe. Users will need to be aware of what options they set and what implications those have: consider:
```
qc = QuickCompress()
sim.operations.add(qc)
while (not qc.complete):
    sim.run(100)

while (sim.timestep < 1e6):
    run(10000)
    print_status_message()

hoomd.dump.GSD.write('filename.gsd')

with gsd.hoomd.open('filename'gsd') as f:
    results = analyze_configuration(f)
    results.write_to_job_dir(job)
```
To make this type of logic work with a dry run, the user will need to place `if dry_run:` logic throughout the script so that the while loops will complete early and that the expected output file is not read or results written to signac jobs. We cannot reasonably expect that gsd and signac will somehow detect that they are related to a HOOMD Simulation object with a dry run flag set and appropriately mock their behavior.

For the portions of file output in HOOMD that we can control, how would you suggest we communicate the dry run options to the classes? Also, who would like to implement this functionality?

**Comment by mphoward** [comment]
I totally agree, examples like this are not suitable for using this option. This is one reason I think it would be better to make a “dry run” just override and shorten the run commands, and still allow the script to execute otherwise as usual. This could be done with only a couple lines, as opposed to needing to forward this option to other parts of the code to control their output. With the intent of this option clear, the user could decide if it makes sense to use it. I think it would cover a lot of “standard” simulation-only workflows, but would not suitable for ones that combine more analysis, etc. in a script.

**Comment by vyasr** [comment]
@mphoward's comment showed up as I was writing this :) I agree with what he said, here's my take.

I agree with you @joaander, that's why I specifically mentioned that users have to handle the dry run for non-HOOMD parts of the code:

> 1. No files would be created by HOOMD commands, but users would be responsible for handling this appropriately if they run anything else that might affect output (e.g. a `signac` job document modification, or with respect to #748, `duecredit` citation generation).

I still would consider this an improvement, because right now users are either commenting out those parts of the code or manually removing/editing files to undo the changes. This isn't all the different from enabling restartable jobs. Even though HOOMD supports them nicely, if you want to use some other tools you need to add logic to check progress in the context of those tools (for instance, `job.doc` queries in `signac`).

That being said, are there examples where we couldn't make this idea work for a pure HOOMD script? Aside from cases where not running long enough actually violates the physics and causing something downstream to explode (for instance, poor equilibration leading to blow up in a future run)? Those limitations seem more critical to me for us to identify since that indicates that the HOOMD API itself cannot manage dry runs flawlessly.
